Bugfix/correct keep alive interval on connect

chillymattster opened the pull request:
Hello,

since the last PR #4 in December 2022 I was quite happy to be able to have a MQTT client on my Wago controllers.
However, I wondered why it took ages for the broker to detect ungraceful connection lost until I recently noticed, that the debug log of mosquitto (hosted via Home Assistant) stated a very big keep alive interval.
Log looks like
`New client connected from 178.179.1.10:53042 as Wago750 (p2, c1, k60000, u'wago')`
The parameter `k60000` indicates an interval of 60.000 seconds (16 hours) although I didn't change it and the default is 1 minute.

Diving into it, I found a conversion of TIME_TO_WORD in AC_SendConnect. In my opinion this is a bug, as the CoDeSys manual states that TIME is based in milliseconds stored in a DWORD.

Therefore I'm happy to provide and discuss this pull request changing the conversion to TIME_TO_DWORD and dividing the value by 1000 to get seconds instead of milliseconds. In addition I added a limitation of keep alive interval to 65535 seconds as well as removing raising an error and jumping to an unknown state in AC_KeepAlive in favor of the same limitation if keep alive interval limit is exceeded.

I tested on my Wago 750-830 and 750-881 controllers. Broker now detects connection lost as expected and publishes last will message after keep alive interval x 1,5 🙂

Please let me know if there is anything else to do for this pull request to be accepted, I'm not a professional PFC developer. 🙃😇

Cheers,
Matthias

Commit:
Fixed serialization of i_tKeepAliveTimeInterval in AC_SendCONNECT

Commit:
Removed error when keep alive internal greater than allowed, just set to max value

huwylphi commented on mqtt.EXP:
```diff
@@ -948,8 +945,13 @@ iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := 0
-iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := WORD_TO_BYTE(SHR(TIME_TO_WORD(i_tKeepAliveTimeInterval), 8));	(* Keep Alive MSB *)
-iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := WORD_TO_BYTE(TIME_TO_WORD(i_tKeepAliveTimeInterval));			(* Keep Alive LSB *)
+(* Valid Keep Alive is an integer from 0 to 65535, representing the maximum time in seconds allowed to elapse between MQTT protocol packets sent by the client. *)
+IF i_tKeepAliveTimeInterval > t#65535s THEN
```
```suggestion
```

huwylphi commented on mqtt.EXP:
```diff
@@ -948,8 +945,13 @@ iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := 0
-iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := WORD_TO_BYTE(TIME_TO_WORD(i_tKeepAliveTimeInterval));			(* Keep Alive LSB *)
+(* Valid Keep Alive is an integer from 0 to 65535, representing the maximum time in seconds allowed to elapse between MQTT protocol packets sent by the client. *)
+IF i_tKeepAliveTimeInterval > t#65535s THEN
+	i_tKeepAliveTimeInterval := t#65535s;
```
```suggestion
```

huwylphi commented on mqtt.EXP:
```diff
@@ -948,8 +945,13 @@ iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := 0
+(* Valid Keep Alive is an integer from 0 to 65535, representing the maximum time in seconds allowed to elapse between MQTT protocol packets sent by the client. *)
+IF i_tKeepAliveTimeInterval > t#65535s THEN
+	i_tKeepAliveTimeInterval := t#65535s;
+END_IF
```
```suggestion
```

huwylphi commented on mqtt.EXP:
```diff
@@ -948,8 +945,13 @@ iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := 0
+	i_tKeepAliveTimeInterval := t#65535s;
+END_IF
+(* Datatype TIME is internally based on milliseconds stored on a DWORD, so TIME_TO_DWORD gets milliseconds. Divide by 1000 to provide a valid protocoll value based on seconds. *)
+iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := DWORD_TO_BYTE(SHR((TIME_TO_DWORD(i_tKeepAliveTimeInterval) / 1000), 8));	(* Keep Alive MSB *)
```
```suggestion
iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := DWORD_TO_BYTE(SHR((TIME_TO_DWORD(MIN(i_tKeepAliveTimeInterval, 65535)) / 1000), 8));	(* Keep Alive MSB *)

```

huwylphi commented on mqtt.EXP:
```diff
@@ -948,8 +945,13 @@ iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := 0
+END_IF
+(* Datatype TIME is internally based on milliseconds stored on a DWORD, so TIME_TO_DWORD gets milliseconds. Divide by 1000 to provide a valid protocoll value based on seconds. *)
+iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := DWORD_TO_BYTE(SHR((TIME_TO_DWORD(i_tKeepAliveTimeInterval) / 1000), 8));	(* Keep Alive MSB *)
+iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := DWORD_TO_BYTE(TIME_TO_DWORD(i_tKeepAliveTimeInterval) / 1000);			(* Keep Alive LSB *)		(* Keep Alive LSB *)
```
```suggestion
iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := DWORD_TO_BYTE(TIME_TO_DWORD(MIN(i_tKeepAliveTimeInterval, 65535)) / 1000);			(* Keep Alive LSB *)

```

huwylphi reviewed (commented):
Hi @chillymattster,

Thank you for your nice analysis and findings!
I would agree with your comments and proposition. The only suggestion I would make is some code simplification by replacing the IF statement for checking the upper limit value by using the native MIN() function.
So basically we could remove
`IF i_tKeepAliveTimeInterval > t#65535s THEN
	i_tKeepAliveTimeInterval := t#65535s;
END_IF`
and introduce the MIN() method here:
`iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := DWORD_TO_BYTE(SHR((TIME_TO_DWORD(MIN(i_tKeepAliveTimeInterval, 65535)) / 1000), 8));	(* Keep Alive MSB *)
iSendIndexCount := iSendIndexCount + 1;	abOutBuffer[iSendIndexCount] := DWORD_TO_BYTE(TIME_TO_DWORD(MIN(i_tKeepAliveTimeInterval, 65535)) / 1000);			(* Keep Alive LSB *)`

(also remove the duplicate comment _(* Keep Alive LSB *))_

I've added those suggestions directly in your PR, so one could be able to use them

Also we should create a new .lib file containing these changes and a new release version with some release notes to keep some history.

Kind regards
Philippe

chillymattster committed:
Apply suggestions from huwylphi, replace IF statements by MIN-function

Co-authored-by: Philippe Huwyler <huwylphi@users.noreply.github.com>

Commit:
Introduced constant for max KeepAlive time

Commit:
Use MIN-function for max KeepAlive time also on AC_KeepAlive

chillymattster commented:
Hi @huwylphi,

thanks for reviewing, very much appreciated 🙂
I like your suggestion and committed it.

In addition I added the same pattern also in AC_KeepAlive. There was another check for not exceeding the max keep alive interval and I replaced it with the MIN-function. As there is more than one using of the max keep alive interval now, I replaced the direct value in the MIN-function and introduced the constant MQTT_KEEPALIVE_MAX_VALUE that is now used. I know it is not likely that this constant will change in the future but I like to have a single point for changing such "hardcoded" values.

Regarding your comment concerning the .lib file and release comments, unfortunately I have no idea how to create or how to contribute on these points. Is this something I can do or @Lugaresi as the repo owner has to do?

Cheers,
Matthias

Commit:
Fixed missing constant

Lugaresi commented:
Hello, 
 Thank you for your continued interest in this project. I have not started using it yet nor maintaining, so I made it where @huwylphi should be able to merge pull requests autonomously. I have quickly looked over the code and it looks good to me. I agree on putting out a new release with its .lib file, if you are interested I will add you to the contributors for this project and if needed amend access privileges where needed.
If not I'll proceed in the next few days.

Regards
 Flavio

huwylphi commented:
Hello @chillymattster and @Lugaresi,
Sorry for being unresponsive for some time...
I'll try to create a new lib file tomorrow and merge it together with that PR.
Anyway, it feels good seeing some people consuming that lib ;)
Have a nice evening and thanks for your contribution.
Philippe

huwylphi commented:
Hi @Lugaresi and @chillymattster,

today I could finally merge this pull request. Since the PR was coming from a forked branch, I decided to not add further changes about updating the lib file to that branch but just merge it. I then created a new branch where I updated the lib file and also merged my changes from my fork, that is: just added one excel file that contains all the possible error messages returned by the lib. After merging that to this repo, I've tagged commits accordingly to the version history and created releases with some change log for each release. See [release section](https://github.com/Lugaresi/codesys2-mqtt-library/releases) in the repo. That way we have some tracking of changes made. I hope this is fine with you.

Have a nice day!
Philippe

chillymattster commented:
Hi @huwylphi;

awesome, many thanks!
